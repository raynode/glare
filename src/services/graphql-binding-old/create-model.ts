import { capitalize, pluralize, singularize } from 'inflection'
import { reduce } from 'lodash'
import { Associations, Attribute, Attributes, BaseModel, Methods, SimpleAttribute, SimpleAttributes } from './types'

export const createAttribute = (attribute: SimpleAttribute): Attribute => ({
  allowNull: true,
  allowUpdate: true,
  comment: '',
  defaultValue: undefined,
  type: 'scalar',
  visible: true,
  ...attribute,
})

export const convertSimpleAttributes = <Type>(simpleAttributes: SimpleAttributes<Type>) =>
  reduce(
    simpleAttributes,
    (memo, simpleAttribute, key) => {
      memo[key] = createAttribute(simpleAttribute)
      return memo
      // tslint:disable-next-line no-object-literal-type-assertion
    },
    {} as Attributes<Type>,
  )

export const createModel = <Type = any>(
  name: string,
  attributes: SimpleAttributes<Type>,
  associations?: Associations<Type>,
): BaseModel<Type> => {
  return {
    associations,
    assocResolvers: [],
    attributes: convertSimpleAttributes(attributes),
    fieldNames: {
      create: `${name}CreateInput`,
      list: `${name}ListType`,
      model: `${name}`,
      order: `${name}OrderEnum`,
      page: `Page`,
      update: `${name}UpdateInput`,
    },
    inspect: () => `${name}Model`,
    name,
    names: {
      create: `onCreate${name}`,
      createOne: `create${name}`,
      delete: `onDelete${name}`,
      deleteMany: `delete${pluralize(name)}`,
      findMany: `${pluralize(name)}`,
      findOne: `${name}`,
      filters: `${name}Filter`,
      updateData: `Update${name}`,
      createData: `Initial${name}`,
      update: `onUpdate${name}`,
      updateOne: `update${name}`,
    },
  }
}
