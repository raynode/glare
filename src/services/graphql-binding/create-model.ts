import { capitalize, pluralize, singularize } from 'inflection'

import { Attributes, BaseModel, Methods } from './types'

export const createModel = <Type = any>(name: string, attributes: Attributes): BaseModel<Type> => {
  return {
    associations: {},
    assocResolvers: [],
    attributes,
    fieldNames: {
      create: `${name}CreateInput`,
      list: `${name}ListType`,
      model: `${name}`,
      order: `${name}OrderEnum`,
      page: `Page`,
      update: `${name}UpdateInput`,
    },
    inspect: () => `${name}Model`,
    name,
    names: {
      create: `onCreate${name}`,
      createOne: `create${name}`,
      delete: `onDelete${name}`,
      deleteMany: `delete${pluralize(name)}`,
      findMany: `${pluralize(name)}`,
      findOne: `${name}`,
      filters: `${name}Filter`,
      updateData: `Update${name}`,
      createData: `Initial${name}`,
      update: `onUpdate${name}`,
      updateOne: `update${name}`,
    },
  }
}
