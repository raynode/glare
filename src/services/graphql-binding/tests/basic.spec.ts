// tslint:disable-next-line
import * as faker from 'faker'

import { createModel } from '../create-model'

import { modelsToSDL } from './test-utils'

describe('basic graphql generation', () => {
  it('should create a basic graphql setup from a model', () => {
    const model = createModel(
      'BasicTest',
      {
        stringField: {
          allowNull: true,
          allowUpdate: true,
          comment: 'this is a test-string field',
          defaultValue: 'none',
          type: 'string',
          visible: true,
        },
      },
      {
        createOne: () => null,
        findOne: () => null,
        deleteMany: () => null,
        findMany: () => null,
        update: () => null,
      },
    )
    expect(modelsToSDL([model])).toMatchSnapshot()
  })
})
