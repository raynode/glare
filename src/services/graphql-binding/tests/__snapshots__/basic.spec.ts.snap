// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic graphql generation should create a basic graphql setup from a model 1`] = `
"type BasicTest {
  stringField: String
}

input BasicTestFilter {
  stringField: String
}

enum BasicTestOrderEnum {
  stringField_ASC
  stringField_DESC
}

input BasicTestUpdateInput {
  stringField: String
}

input InitialBasicTest {
  stringField: String
}

type Mutation {
  createBasicTest(data: InitialBasicTest!): BasicTest
  updateBasicTest(where: BasicTestFilter!, data: BasicTestUpdateInput!): BasicTest
  deleteBasicTests(where: BasicTestFilter!): [BasicTest!]!
}

input Page {
  limit: Int
  offset: Int
}

type Query {
  BasicTest(where: BasicTestFilter!, order: BasicTestOrderEnum): BasicTest
  BasicTests(where: BasicTestFilter, order: BasicTestOrderEnum, page: Page): [BasicTest!]!
}
"
`;
